File: simulator/speed_limit_lookup.py
"""
Tra cuu maxspeed theo vi tri (lat, lng), dua tren du lieu OSM da tai san
tu prepare_road_speed_data.py. Load 1 lan luc import, KHONG goi network.
"""

import json
import math
from pathlib import Path

_DATA_PATH = Path(__file__).parent / "data" / "hanoi_road_speeds.json"

EARTH_RADIUS_KM = 6371.0

# Danh sach segment: (lat1, lng1, lat2, lng2, maxspeed) - dung cho tra cuu
_segments: list[tuple[float, float, float, float, int]] = []


def _load():
    global _segments
    if not _DATA_PATH.exists():
        print(
            f"[speed_limit_lookup] KHONG tim thay {_DATA_PATH} - "
            f"chay prepare_road_speed_data.py truoc. Fallback ve maxspeed=40 cho toan bo."
        )
        return
    with open(_DATA_PATH, "r", encoding="utf-8") as f:
        data = json.load(f)
    segs = []
    for way in data["ways"]:
        coords = way["coords"]
        for i in range(len(coords) - 1):
            lat1, lng1 = coords[i]
            lat2, lng2 = coords[i + 1]
            segs.append((lat1, lng1, lat2, lng2, way["maxspeed"]))
    _segments = segs
    print(f"[speed_limit_lookup] Da load {len(_segments)} segment duong.")


def _point_to_segment_km(lat, lng, lat1, lng1, lat2, lng2, ref_lat) -> float:
    """Khoang cach xap xi (km) tu diem den doan thang, dung phep chieu
    phang (equirectangular) - du chinh xac cho pham vi bbox nho (~35km),
    nhanh hon nhieu so voi tinh haversine chinh xac cho tung diem tren doan."""
    km_per_deg_lat = 111.0
    km_per_deg_lng = 111.0 * math.cos(math.radians(ref_lat))

    x, y = lng * km_per_deg_lng, lat * km_per_deg_lat
    x1, y1 = lng1 * km_per_deg_lng, lat1 * km_per_deg_lat
    x2, y2 = lng2 * km_per_deg_lng, lat2 * km_per_deg_lat

    dx, dy = x2 - x1, y2 - y1
    if dx == 0 and dy == 0:
        return math.hypot(x - x1, y - y1)

    t = max(0.0, min(1.0, ((x - x1) * dx + (y - y1) * dy) / (dx * dx + dy * dy)))
    proj_x, proj_y = x1 + t * dx, y1 + t * dy
    return math.hypot(x - proj_x, y - proj_y)


MAX_MATCH_DISTANCE_KM = 0.05  # 50m - qua nguong nay coi nhu khong co duong lon nao gan
RESIDENTIAL_FALLBACK_SPEED = 30


def get_speed_limit(lat: float, lng: float) -> int:
    if not _segments:
        return 40

    best_dist = float("inf")
    best_speed = RESIDENTIAL_FALLBACK_SPEED
    for lat1, lng1, lat2, lng2, speed in _segments:
        d = _point_to_segment_km(lat, lng, lat1, lng1, lat2, lng2, lat)
        if d < best_dist:
            best_dist = d
            best_speed = speed

    if best_dist > MAX_MATCH_DISTANCE_KM:
        return RESIDENTIAL_FALLBACK_SPEED
    return best_speed


_load()

File: backend/src/services/speedLimitLookup.js
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

// Port 1:1 tu simulator/speed_limit_lookup.py (Python) sang JS, dung
// chung file du lieu hanoi_road_speeds.json (khong lien network, load
// 1 lan luc module duoc import). Ly do port lai thay vi goi qua
// child_process: route nay bi goi moi lan telemetry (~8s/lan/trip dang
// chay) - spawn process moi lan se cong don latency khong dang.
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DATA_PATH = path.join(__dirname, '..', 'data', 'hanoi_road_speeds.json');

const MAX_MATCH_DISTANCE_KM = 0.05; // 50m - qua nguong nay coi nhu khong co duong lon nao gan
const RESIDENTIAL_FALLBACK_SPEED = 30;

// Danh sach segment dang [lat1, lng1, lat2, lng2, maxspeed]
let segments = [];

function load() {
    try {
        const raw = readFileSync(DATA_PATH, 'utf-8');
        const data = JSON.parse(raw);
        const segs = [];
        for (const way of data.ways) {
            const coords = way.coords;
            for (let i = 0; i < coords.length - 1; i++) {
                const [lat1, lng1] = coords[i];
                const [lat2, lng2] = coords[i + 1];
                segs.push([lat1, lng1, lat2, lng2, way.maxspeed]);
            }
        }
        segments = segs;
        console.log(`[speedLimitLookup] Da load ${segments.length} segment duong.`);
    } catch (err) {
        console.error(
            `[speedLimitLookup] KHONG doc duoc ${DATA_PATH} - fallback maxspeed=40 cho toan bo.`,
            err.message,
        );
    }
}

// Khoang cach xap xi (km) tu diem den doan thang, dung phep chieu phang
// (equirectangular) - du chinh xac cho pham vi bbox nho (~35km), y het
// cong thuc ben Python de dam bao 2 phia cho ra cung ket qua.
function pointToSegmentKm(lat, lng, lat1, lng1, lat2, lng2, refLat) {
    const kmPerDegLat = 111.0;
    const kmPerDegLng = 111.0 * Math.cos((refLat * Math.PI) / 180);

    const x = lng * kmPerDegLng;
    const y = lat * kmPerDegLat;
    const x1 = lng1 * kmPerDegLng;
    const y1 = lat1 * kmPerDegLat;
    const x2 = lng2 * kmPerDegLng;
    const y2 = lat2 * kmPerDegLat;

    const dx = x2 - x1;
    const dy = y2 - y1;
    if (dx === 0 && dy === 0) {
        return Math.hypot(x - x1, y - y1);
    }

    let t = ((x - x1) * dx + (y - y1) * dy) / (dx * dx + dy * dy);
    t = Math.max(0, Math.min(1, t));
    const projX = x1 + t * dx;
    const projY = y1 + t * dy;
    return Math.hypot(x - projX, y - projY);
}

/**
 * Tra cuu maxspeed (km/h) gan vi tri (lat, lng) nhat. Brute-force qua
 * ~46.7k segment (~7965 way) - da do thu tren Node, du nhanh cho tan
 * suat goi 1 lan/8s/trip (khong can spatial index cho scope do an).
 */
export function getSpeedLimit(lat, lng) {
    if (segments.length === 0) return 40;

    let bestDist = Infinity;
    let bestSpeed = RESIDENTIAL_FALLBACK_SPEED;
    for (const [lat1, lng1, lat2, lng2, speed] of segments) {
        const d = pointToSegmentKm(lat, lng, lat1, lng1, lat2, lng2, lat);
        if (d < bestDist) {
            bestDist = d;
            bestSpeed = speed;
        }
    }

    if (bestDist > MAX_MATCH_DISTANCE_KM) return RESIDENTIAL_FALLBACK_SPEED;
    return bestSpeed;
}

load();
